lib/sshutils/networking/networking.go

```diff
@@ -0,0 +1,341 @@
+	return nil
+}
+
+// Dial requests a network connection from the networking subprocess.
+func (p *Process) Dial(ctx context.Context, network string, addr string) (net.Conn, error) {
+	connFD, err := p.request(ctx, Request{
+		Operation: NetworkingOperationDial,
+		Network:   network,
+		Address:   addr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer connFD.Close()
+
+	conn, err := net.FileConn(connFD)
+	return conn, trace.Wrap(err)
+}
+
+// Listen requests a local listener from the networking subprocess.
+func (p *Process) Listen(ctx context.Context, network string, addr string) (net.Listener, error) {
+	listenerFD, err := p.request(ctx, Request{
+		Operation: NetworkingOperationListen,
+		Network:   network,
+		Address:   addr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer listenerFD.Close()
+
+	listener, err := net.FileListener(listenerFD)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if unixListener, ok := listener.(*net.UnixListener); ok {
+		unixListener.SetUnlinkOnClose(true)
+	}
+
+	return listener, nil
+}
+
+// ListenAgent requests a local ssh-agent listener from the networking subprocess.
+func (p *Process) ListenAgent(ctx context.Context) (net.Listener, error) {
+	listenerFD, err := p.request(ctx, Request{
+		Operation: NetworkingOperationListenAgent,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer listenerFD.Close()
+
+	listener, err := net.FileListener(listenerFD)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	unixListener, ok := listener.(*net.UnixListener)
+	if !ok {
+		return nil, trace.BadParameter("expected *net.UnixListener but got %T", listener)
+	}
+
+	unixListener.SetUnlinkOnClose(true)
+	return unixListener, nil
+}
+
+// ListenX11 requests a local XServer listener from the networking subprocess.
+func (p *Process) ListenX11(ctx context.Context, req X11Request) (*net.UnixListener, error) {
+	listenerFD, err := p.request(ctx, Request{
+		Operation:  NetworkingOperationListenX11,
+		X11Request: req,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer listenerFD.Close()
+
+	listener, err := net.FileListener(listenerFD)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	unixListener, ok := listener.(*net.UnixListener)
+	if !ok {
+		return nil, trace.BadParameter("expected *net.UnixListener but got %T", listener)
+	}
+
+	unixListener.SetUnlinkOnClose(true)
+	return unixListener, nil
+}
+
+// request sends a networking request to the networking process and waits
+// for a file corresponding to an open network connection or listener.
+func (p *Process) request(ctx context.Context, req Request) (*os.File, error) {
+	ctx, cancel := context.WithTimeout(ctx, defaults.DefaultIOTimeout)
+	defer cancel()
+
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	requestConn, err := p.newConnection()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	context.AfterFunc(ctx, func() { requestConn.Close() })
```

Comment:
```suggestion
	defer context.AfterFunc(ctx, func() { requestConn.Close() })()
```

Reply:
Except that now the request connection is datagram so closing it is no longer a signal for the child process, so the context in any of the operations here is a lie.

Reply:
I changed it to a stream so that closing it would be reflected in the child process, rather than managing an extra fd for closure. I can revert https://github.com/gravitational/teleport/pull/43756/commits/dc76a7bb3f22f49baef45d59775eb0b7c4614405

Reply:
Reverted to stream socket. Why would this need to be deferred? we want to unblock read on the request conn if the context is closed.

Reply:
I ended up having to remove this connection-context logic anyways due to it causing an x11 test to be flaky - [42095d6](https://github.com/gravitational/teleport/pull/43756/commits/42095d6e49cbaca8ed4c1595ba61848d11cc25a7). I couldn't figure out what the root cause of the flake was, some test environment fishiness probably. Once that's sorted we could revert [42095d6](https://github.com/gravitational/teleport/pull/43756/commits/42095d6e49cbaca8ed4c1595ba61848d11cc25a7) back.

Reply:
Nevermind I found the culprit b9062c73e0dda3793eecf94b722007b7cade9598